### PR TITLE
SEQNG-32 Implemented breakpoints in execution engine

### PR DESCRIPTION
**Pull request by jluhrs**

### CODE REVIEW

File: modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/package.scala [hunk 1]
```diff
@@ -91,18 +102,18 @@ package object engine {
               // Final State
               case Some(qs: Sequence.State.Final) =>
                 putS(id)(qs) *> send(q)(finished(id))
-              // Execution completed
+              // Execution completed. Check breakpoint here
               case Some(qs) =>
-                putS(id)(qs) *> send(q)(executing(id))
+                putS(id)(qs) *> (if(qs.getCurrentBreakpoint) switch(q)(id)(SequenceState.Idle)
```
Comment (jluhrs) [inline modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/package.scala]: I'm sure there is a more elegant way of writing this.

File: modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Step.scala [hunk 2]
```diff
@@ -8,10 +8,12 @@ import Scalaz._
 /**
   * A list of `Executions` grouped by observation.
   */
-case class Step[+A](id: Int, config: StepConfig, executions: List[List[A]])
+case class Step[+A](id: Step.Id, config: StepConfig, breakpoint: Boolean, executions: List[List[A]])
 
 object Step {
 
+  type Id = Int
```
Comment (cquiroz) [inline modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Step.scala]: Is this the position? What happens if we reload the sequence and the steps have changed?

File: modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/SeqexecEngine.scala [hunk 3]
```diff
@@ -74,6 +75,7 @@ class SeqexecEngine(settings: SeqexecEngine.Settings) {
       case engine.Start(_)    => SequenceStart(svs)
       case engine.Pause(_)    => SequencePauseRequested(svs)
       case engine.Load(_, _)  => SequenceLoaded(svs)
+      case engine.Breakpoint(_, _, _) => StepBreakpointChanged(svs)
```
Comment (cquiroz) [inline modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/SeqexecEngine.scala]: Would you align the `=>`

File: modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Sequence.scala [hunk 4]
```diff
@@ -177,33 +180,9 @@ object Sequence {
       * Initialize a `State` passing a `Queue` of pending `Sequence`s.
       */
     // TODO: Make this function `apply`?
-    def init(q: Sequence[Action]): State = Initial(q, SequenceState.Idle)
+    def init(q: Sequence[Action]): State = Sequence.Zipper.currentify(q).map(Zipper(_, SequenceState.Idle)).getOrElse(Final(Sequence.empty(q.id), SequenceState.Idle))
```
Comment (jdnavarro) [inline modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Sequence.scala]: Could this line be broken in 2?

File: modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/package.scala (hunk 1 above)
Comment (jdnavarro) [inline modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/package.scala]: You could create analogous function to `getCurrentBreakPoint` lifted to `Handle`, like `status` in this module I think, but if this only place where you do this, I don't think it's worth the effort.

File: modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Step.scala (hunk 2 above)
Comment (jluhrs) [inline modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Step.scala]: For the time being the id is generated internally using the step position. If the sequence is reloaded its internal representation is regenerated from scratch. Therefore, the breakpoints are lost.

File: modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/Model.scala [hunk 5]
```diff
@@ -95,6 +95,7 @@ object Model {
   object SequenceState {
     case object Completed         extends SequenceState
     case object Running           extends SequenceState
+    case object Stopping          extends SequenceState
```
Comment (jdnavarro) [inline modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/Model.scala]: 😕 We had some discussions while you were away about banning the word `Stop` from the code and rename it all to `Pause`. In any case if this doesn't show up in the UI, I think we'll be fine.

**Pull request merged**